src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -809,7 +809,23 @@ DagState uniqifyState(ImmutableArray<StateForCase> cases, ImmutableDictionary<Bo
         /// </summary>
         private void ComputeBoundDecisionDagNodes(DecisionDag decisionDag, BoundLeafDecisionDagNode defaultDecision)
         {
-            RoslynDebug.Assert(_defaultLabel != null);
+            Debug.Assert(_defaultLabel != null);
+            Debug.Assert(defaultDecision != null);
+
+            // Process the states in topological order, leaves first, and assign a BoundDecisionDag to each DagState.
+            bool wasAcyclic = decisionDag.TryGetTopologicallySortedReachableStates(out ImmutableArray<DagState> sortedStates);
+            if (!wasAcyclic)
+            {
+                // If the dag contains a cycle, return a short-circuit dag instead.
+                decisionDag.RootNode.Dag = defaultDecision;
+
+                // Since we intended the set of DagState nodes to be acyclic by construction, we do not know how
+                // this can happen, but since it does occasionally happen we need to recover gracefully to avoid
+                // crashing the compiler.  See also https://github.com/dotnet/roslyn/issues/45946
+                // If you figure out how it occurs, please modify the DagState construction process to avoid that.
```

Comment:
Should this comment be updated? Perhaps: "Since we intended the set of DagState nodes to be acyclic by construction this is not expected. If the assert fails, please modify the DagState construction process to avoid the cycle."